src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs

```diff
@@ -98,13 +122,39 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
         {
             Interop.SspiCli.ContextFlags unusedAttributes = default;
 
-            InputSecurityBuffers inputBuffers = default;
+            scoped InputSecurityBuffers inputBuffers = default;
             inputBuffers.SetNextBuffer(new InputSecurityBuffer(inputBuffer, SecurityBufferType.SECBUFFER_TOKEN));
             inputBuffers.SetNextBuffer(new InputSecurityBuffer(default, SecurityBufferType.SECBUFFER_EMPTY));
             if (context == null && sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0)
             {
-                byte[] alpnBytes = ConvertAlpnProtocolListToByteArray(sslAuthenticationOptions.ApplicationProtocols);
-                inputBuffers.SetNextBuffer(new InputSecurityBuffer(new ReadOnlySpan<byte>(alpnBytes), SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                var alpn = sslAuthenticationOptions.ApplicationProtocols;
+
+                if (alpn.Count == 1 && alpn[0] == SslApplicationProtocol.Http11)
+                {
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(s_http1, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                }
+                else if (alpn.Count == 1 && alpn[0] == SslApplicationProtocol.Http2)
+                {
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(s_http2, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                }
+                else if (alpn.Count == 2 && alpn[0] == SslApplicationProtocol.Http11 && alpn[1] == SslApplicationProtocol.Http2)
+                {
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(s_http12, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+
+                }
+                else if (alpn.Count == 2 && alpn[0] == SslApplicationProtocol.Http2 && alpn[1] == SslApplicationProtocol.Http11)
+                {
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(s_http21, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                }
+                else
+                {
+                    int protocolLenngth = Interop.Sec_Application_Protocols.GetProtocolLength(alpn);
+                    int bufferLength = sizeof(Interop.Sec_Application_Protocols) + protocolLenngth;
+                    scoped Span<byte> alpnBuffer = bufferLength <= 64 ? stackalloc byte[bufferLength] : new byte[bufferLength];
+
+                    Interop.Sec_Application_Protocols.SetProtocols(alpnBuffer, alpn, protocolLenngth);
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(alpnBuffer, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                }
```

Comment:
Is this if/else chain an exact duplicate of the one earlier in the file?  Can it be extracted into a shared helper?

Reply:
updated into helper function.